flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/AlterTableSchemaUtil.java

```diff
@@ -0,0 +1,525 @@
+            }
+        }
+
+        private void populateWatermarkFromSourceTable(Schema sourceSchema) {
+            for (Schema.UnresolvedWatermarkSpec sourceWatermarkSpec :
+                    sourceSchema.getWatermarkSpecs()) {
+                watermarkSpecs.put(sourceWatermarkSpec.getColumnName(), sourceWatermarkSpec);
+            }
+        }
+
+        private void updateColumn(List<SqlNode> alterColumnPositions) {
+            collectColumnPosition(alterColumnPositions);
+            for (SqlNode alterColumnPos : alterColumnPositions) {
+                SqlTableColumn alterColumn = ((SqlTableColumnPosition) alterColumnPos).getColumn();
+                if (alterColumn instanceof SqlTableColumn.SqlComputedColumn) {
+                    convertComputedColumn((SqlTableColumn.SqlComputedColumn) alterColumn);
+                } else {
+                    convertNonComputedColumn(alterColumn);
+                }
+            }
+        }
+
+        private void updatePrimaryKey(SqlTableConstraint alterPrimaryKey) {
+            checkPrimaryKeyExists();
+            List<String> primaryKeyColumns = Arrays.asList(alterPrimaryKey.getColumnNames());
+            if (alterColNames.isEmpty()) {
+                primaryKeyColumns.forEach(this::updatePrimaryKeyNullability);
+            }
+            primaryKey =
+                    new Schema.UnresolvedPrimaryKey(
+                            alterPrimaryKey
+                                    .getConstraintName()
+                                    .orElseGet(
+                                            () ->
+                                                    primaryKeyColumns.stream()
+                                                            .collect(
+                                                                    Collectors.joining(
+                                                                            "_", "PK_", ""))),
+                            primaryKeyColumns);
+        }
+
+        private void updatePrimaryKeyNullability(String columnName) {
+            Schema.UnresolvedColumn column = columns.get(columnName);
+            if (column instanceof Schema.UnresolvedPhysicalColumn) {
+                AbstractDataType<?> originType =
+                        ((Schema.UnresolvedPhysicalColumn) column).getDataType();
+                columns.put(
+                        columnName,
+                        new Schema.UnresolvedPhysicalColumn(
+                                columnName,
+                                originType.notNull(),
+                                column.getComment().orElse(null)));
+            }
+        }
+
+        private void updateWatermark(SqlWatermark alterWatermarkSpec) {
+            checkWatermarkExists();
+            SqlIdentifier eventTimeColumnName = alterWatermarkSpec.getEventTimeColumnName();
+            String rowtimeField = String.join(".", eventTimeColumnName.names);
+            watermarkSpecs.clear();
```

Comment:
> Why clear? I think overwrite is okay?
Because the rowtime field might be changed. Consider the following case
```sql
CREATE TABLE T (
  ts1 TIMESTAMP(3),
  ts2 TIMESTAMP(3),
  order_id BIGINT,
  WATERMARK FOR ts1 AS ts1 - INTERVAL '1' DAY
);

ALTER TABLE T MODIFY WATERMARK FOR ts2 AS ts2;
```